File: 1_Introduction/Memory_Access_And_Caches/go_code/caching.go
package main

import (
	"fmt"
	"time"
)

const ITERS = 20000

type f func()

func v1() {
	x := [ITERS][ITERS]int{}
	var i, j int
	for i = 0; i < ITERS; i++ {
		for j = 0; j < ITERS; j++ {
			x[j][i] = i + j
		}
	}
}

func v2() {
	x := [ITERS][ITERS]int{}
	var i, j int
	for j = 0; j < ITERS; j++ {
		for i = 0; i < ITERS; i++ {
			x[j][i] = i + j
		}
	}
}

func printSpeed(fn f, version string) {
	start := time.Now()
	fn()
	end := time.Now()
	elapsed := end.Sub(start).Seconds()
	fmt.Printf("%s time: %.3f\n", version, elapsed)
}

func main() {
	printSpeed(v1, "v1")
	printSpeed(v2, "v2")
}

File: 1_Introduction/Memory_Access_And_Caches/py_code/caching.py
import time

ITERS = 4000


x = [[i for i in range(ITERS)] for j in range(ITERS)]


def v1():
    for i in range(ITERS):
        for j in range(ITERS):
            x[j][i] = i + j


def v2():
    for j in range(ITERS):
        for i in range(ITERS):
            x[j][i] = i + j


def print_time(f, version):
    start = time.time()
    f()
    end = time.time()
    print("{} Time: {:.1f}".format(version, end - start))


def main():
    print_time(v1, "v1")
    print_time(v2, "v2")


main()
